File: algorithms/sorting/insertion-sort/insertion-sort.hs
insert' :: Ord a => a -> [a] -> [a]
insert' e [] = [e]
insert' e xs@(x : xs') =
  case compare e x of
    GT -> x : insert' e xs'
    _ -> e : xs

insertionSort :: Ord a => [a] -> [a]
insertionSort = foldr insert' []

myList :: [Integer]
myList = [9, 5, 6, 3, 1, 8, 4, 7, 2]

main :: IO ()
main = do
  putStrLn "Input:"
  print myList
  putStrLn ""
  putStrLn "Output:"
  print (insertionSort myList)

File: algorithms/sorting/insertion-sort/insertion-sort.cpp
#include <iostream>
using namespace std;

void insertionSort(int arr[], int n) {
  if (n < 2) return;
  for (int i = 1; i < n; ++i) {
    int c = arr[i];
    int p = i - 1;
    while (p >= 0 && arr[p] > c) {
      swap(arr[p], arr[p + 1]);
      --p;
    }
  }
}

int main() {
  int arr[] = {9, 8, 7, 6, 5, 4, 3, 2, 1, 0};
  int length = sizeof(arr) / sizeof(int);
  cout << "Input: " << endl;
  for (int x : arr) {
    cout << x << ", ";
  }
  insertionSort(arr, length);
  cout << endl << endl << "Output: " << endl;
  for (int x : arr) {
    cout << x << ", ";
  }
  cout << endl;

  return 0;
}